src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/Protocol/ApplyWorkspaceEditParams.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```

Comment:
all our own versions of these lsp objects.  

pros: 

1. follows the same approach as MS.VS.LS.Protocol.  
2. Is binary compatible for our code (modulo updating our namespace).
3. Can be tweaked as we see fit on our own schedule

Cons:

1. Newtonsoft based.
2. May diverge from VS as they add support for things.  Will require manual upkeep.
3. Large.

Reply:
> Can be tweaked as we see fit on our own schedule

Razor has the same problem to solve with MS.VS.LS.Protocol. Would you rather us not use these?

Reply:
We need to do something here in regards to clasp / this with Razor.  currently we basically cannot make any changes to clasp (of which I've wanted to make quite a fiew) without doing a dual insertion.  I would like us to avoid more places where we need that, especially because LSP compatibility *does not mean binary compat*

Reply:
> without doing a dual insertion

I would block Razor using these types until we figure out how / if we can share it without dual insertions

I think we should also strongly consider using Karthik / Matteo's generation library for these (though doesn't have to be now).  If we're using them,

Reply:
yeah.  we could potenitally make this a source package or something as well.  it truly is just serialization goo. 

does Karthik's lib handle all the extension cases?

Reply:
@dibarbet does this change require a dual insertion? It seems to me like it would be binary compatible, but you said "any changes to clasp" so I want to be sure.

Reply:
This change itself doesn't (unless existing clasp APIs are being changed).  But as soon as Razor started using types from here we couldn't make breaking changes to the existing definitions (which can happen in LSP, for example adding new sum types).

There's a separate discussion on how to handle breaking changes in clasp which I'm looking into which may mitigate this.

Reply:
Though - I would also avoid putting these definitions in clasp.  It should probably be a different (source-only) package.

Reply:
would having VS platform owned source package solve the issue we are tying to solve?

Reply:
@olegtk yes. If the license for those files didn't conflict in some way.

Reply:
wouldn't it cause PerfDDRIT regression though?

Reply:
Not as a source package. It wouldn't make a new dll. It would literally just me code added to Roslyn.  Just maintained in one location.